lib/Parse/ParseExpr.cpp

```diff
@@ -2110,6 +2113,56 @@ ParserResult<Expr> Parser::parseExprStringLiteral() {
                                       AppendingExpr));
 }
 
+ParserResult<Expr> Parser::parseExprRegexLiteral() {
+  assert(Tok.is(tok::regex_literal));
+  SyntaxParsingContext LocalContext(SyntaxContext,
+                                    SyntaxKind::RegexLiteralExpr);
+
+  auto *parseRegexFn = Lexer::getExperimentalRegexParsingFn();
+  assert(parseRegexFn);
+
+  // Strip off delimiters.
+  auto rawText = Tok.getText();
+  assert(rawText.front() == '\'' && rawText.back() == '\'');
+  auto regexText = rawText.slice(1, rawText.size() - 1);
+
+  // Let the Swift library parse the contents, returning an error, or null if
+  // successful.
+  // TODO: We need to be able to pass back a source location to emit the error
+  // at.
+  auto *errorStr = parseRegexFn(regexText.str().c_str());
+  if (errorStr)
+    diagnose(Tok, diag::regex_literal_parsing_error, errorStr);
+
+  auto loc = consumeToken();
+
+  // Create an implicit .init(_regexString: "<regex text>") call to serve as the
+  // semantic expression for the regex. The type-checker will provide it with
+  // the correct contextual type. We delay the contextual type for a couple of
```

Comment:
Can we move this logic to `Sema` (e.g. `PreCheckExpression`)? In general, we want to move semantic logic out of `Parser` as much as possible.
I think `RegexExpr::createParsed` can initialize `SemanticExpr` with `nullptr`

Reply:
I'd rather not introduce mutability to the AST, but I guess this synthesis could be done lazily (maybe through a request)?

Reply:
@rxwei is planning on removing this in a follow-up (in favor of logic in CSApply + SILGen), so I'm going to leave it as-is for now